relevanceai/vector_tools/plot_txt_gist.py

```diff
@@ -0,0 +1,202 @@
+        """
+        * Lower-casing
+        * Digit removal
+        * Punctuation removal
+        """
+        if lower:
+            txt = txt.lower()
+        if remove_digit:
+            txt = ''.join([ch for ch in txt if ch not in string.digits])
+        if remove_punct:
+            txt = ''.join([ch for ch in txt if ch not in string.punctuation])
+        return txt
+
+    @staticmethod
+    def get_word_frequency(
+        str_list:str = List[str], 
+        remove_stop_words: bool = True, 
+        additional_stop_words: List[str] = [])->List:
+        if remove_stop_words:
+            stpw = stopwords.words('english')
+            stpw += additional_stop_words
+        else:
+            stpw = []
+        word_counter = Counter([w.lower() for s in str_list for w in s.split() if w not in stpw])
+        return sorted(word_counter.items(), key=lambda item: (-item[1], item[0]))
+
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+from sklearn import preprocessing
+from tqdm.auto import tqdm
+from typing import List
+from relevanceai.api.client import BatchAPIClient
+from relevanceai.vector_tools.cluster import KMeans
+#ToDo: DocUtils reference to normalize_text, get_word_frequency
+from relevanceai.vector_tools.dim_reduction import Ivis
+#from ivis import Ivis
+
+class plot_textgist_model(BatchAPIClient, base_text_processing):  #ToDo: remove base_text_processing
+    def __init__(
+        self,
+        client,        # ToDo: How to use it via BatchAPIClient
+        # dataset_info
+        dataset_id:str,
+        upload_chunksize : int = 50,
+        # clustering
+        cluster_field : str = "_cluster_",
+        #dimensionality reduction parameters
+        embedding_dims=2,
+        dim_red_k: int = 800,
+        n_epochs_without_progress=100
+    ):
+        self.client = client
+        self.dataset_id = dataset_id
+        self.upload_chunksize = upload_chunksize
+        self.cluster_field = cluster_field
+        self.embedding_dims = embedding_dims
+        self.dim_red_k = dim_red_k
+        self.n_epochs_without_progress = n_epochs_without_progress
+
+    def build_and_plot_clusters(
+        self,
+        vector_field : str,
+        text_field : str,
+        max_doc_num : int = None,
+        k : int = 10,
+        alias: str = "kmeans",
+        lower:bool = True,
+        remove_digit:bool = True,
+        remove_punct: bool = True,
+        remove_stop_words: bool = True, 
+        additional_stop_words: List[str] = [],
+        cluster_representative_cnt : int= 3
+        ):
+        # get documents
+        docs = self.get_documents(vector_field = vector_field, text_field = text_field, max_doc_num = max_doc_num)
+
+        # perform kmeans clustering
+        alias = alias + '_' + str(k)
+        centers, clustered_docs = self.kmeans_clustering(docs=docs, vector_field = vector_field, k = k, alias = alias)
+        for i, cl_doc in tqdm(enumerate(clustered_docs)):
```

Comment:
`enumerate(tqdm` or else the progress bar doesn't provide an end value because `enumerate` is an iterable